WWST-7655, added FP for the SAMSUNG LED(Fan lightings)

RaihaPark opened the pull request:
For the ceiling fan lightings, we create a folder "zigbee-ceiling-fan-light.src" in the smartthings directory.
In the test with local public, the device works perfectly.
It contains 4 files : .st-ignore , README.md , led-fan-lightings.groovy , itm-fan-child.groovy

It makes two device cards : lightings and fan

In the lightings card, you can control on / off, brightness level.
In the fan card, you can control on / off, fan speed.

Please review and approve this request.

RaihaPark committed:
Update zigbee-rgbw-bulb.groovy

RaihaPark committed:
Merge branch 'SmartThingsCommunity:master' into master

RaihaPark committed:
Create .st-ignore

2021.06.18 Ceiling fan light added

RaihaPark committed:
Create README.md

2021.06.18 first report

RaihaPark committed:
Create led-fan-lightings.groovy

2021.06.18 create first main lightings file

RaihaPark committed:
Create itm-fan-child.groovy

2021.06.18 first time : create a child device(fan) file

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/.st-ignore:
```diff
@@ -0,0 +1,2 @@
+.st-ignore
```
This file is not needed

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,268 @@
+
+
+metadata {
+    definition(name: "ITM Fan Child", namespace: "SAMSUNG LED", author: "SAMSUNG LED", ocfDeviceType: "oic.d.fan", genericHandler: "Zigbee"){//, mnmn: "SmartThings"){//, vid: "SmartThings-smartthings-Z-Wave_Fan_Controller_4_Speed"){  
```
Remove `genericHandler: "Zigbee"` flag

PKacprowiczS replied:
Remove all out-of-date, commented out code

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,268 @@
+		command "high"
+        command "max"
+		command "raiseFanSpeed"
+		command "lowerFanSpeed"
```
Those commands declaration aren't needed.

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,268 @@
+
+def setRapidCooling(value){
+	log.debug "setRapidCooling $value"
+    if(value == "on")
```
Please fix indentations in both files. Follow code style described here: https://github.com/greens/DTH_development

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,268 @@
+    else{
+        sendEvent(name: "switch", value: "on", displayed: true, isStateChange: true)
+        sendEvent(name: "device.switch", value: "on", displayed: true, isStateChange: true)
+    }   
```
Be persistent in your coding style.
Since mostly you're using
```
def methodName(attribute) {
	<body>
}
```
Keep the same style in all expressions
```
if (condition) {
	<body if true>
} else {
	<body if false>
}
```

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,356 @@
+ *  ZigBee White Color Temperature Bulb
+ *
+ *  Author: SmartThings
+ *  Date: 2015-09-22
```
Dates

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,356 @@
+		max()// + createEvent(name: "fanSpeed", value: 3)
+    } else {
+		max()// + createEvent(name: "fanSpeed", value: 3)
+    }    
```
Instead of defining `low()`, `medium()` ... methods, can't you just call `send_fanSpeed(val)` directly from here? And just add comment describing speed.

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,356 @@
+def installed() {
+	log.debug "Samsung ITM test prod installed"
+    addChildFan()
+    configure() 
```
`configure()` will be called automatically if just add `capability "Configuration"`. There's no need to call it manually

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,356 @@
+	log.debug "Samsung ITM test prod installed"
+    addChildFan()
+    configure() 
+    return
```
This return could be removed

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,268 @@
+    
+    log.trace "[isStateChange] ${eventChange}"
+    return eventChange
+}
```
you don't have to include this method. This will be automatically defined when event is created by Device API. If you want, you can "hardcode" it to an event (so status always will be interpreted as (not) changed) by adding additional flag to an event `isStateChange: _bool_`.

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,268 @@
+def raiseFanSpeed() {
+log.debug "raiseFanSpeed"
+	setFanSpeed(Math.min((device.currentValue("fanSpeed") as Integer) + 1, 4))
+}
```
Won't `device.currentValue("fanSpeed")` always return null, as you actually never set any value to this attribute?

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,268 @@
+    if (speed as Integer == 0) 
+    {
+        sendEvent(name: "switch", value: "off", displayed: true, isStateChange: true)
+        sendEvent(name: "device.switch", value: "off", displayed: true, isStateChange: true)
```
Why are you doubling those events?

PKacprowiczS reviewed (changes requested):
Can you elaborate more on commands/events flow? Cause I've got really confused. 
You define `capability "Fan Speed"` in child DTH and its purpose is to call parent's `send_fanSpeed()`, but why you don't create any "fanSpeed" events? That's probably causing a timeout error. You have 6 child DTHes and you don't differentiate them in any way, when they're calling parent's method (which would be needed if you've tried to create any event in return for childs).

Also code is quite difficult to read, which could affect my ability to properly understand what's going on.

RaihaPark committed:
Merge branch 'SmartThingsCommunity:master' into master

RaihaPark committed:
Delete .st-ignore

20210708 delete

RaihaPark committed:
Update itm-fan-child.groovy

20210728 clean code update

RaihaPark committed:
Update led-fan-lightings.groovy

20210727 clean code update

RaihaPark commented:
We update a new clean code
Please review this again

PKacprowiczS reviewed (changes requested):
when I've asked 
```
Please fix indentations in both files. Follow code style described here: https://github.com/greens/DTH_development
```
I didn't meant one line, I've meant all files. Please adjust your code to the style described in the link above.

Furthermore, please follow this insight from earlier https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/68445#discussion_r659673544

RaihaPark committed:
Update led-fan-lightings.groovy

Edit code style

RaihaPark committed:
Update led-fan-lightings.groovy

Edit code style

RaihaPark committed:
Update itm-fan-child.groovy

Edit code style

RaihaPark committed:
Update led-fan-lightings.groovy

Edit code style

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,249 @@
+import groovy.json.JsonOutput
+
+metadata {
+	definition (name: "LED FAN lightings", namespace: "SAMSUNG LED", author: "SAMSUNG LED", runLocally: true, minHubCoreVersion: '000.019.00012', executeCommandsLocally: true, genericHandler: "Zigbee") {
```
Remove those flags `runLocally: true, minHubCoreVersion: '000.019.00012', executeCommandsLocally: true, genericHandler: "Zigbee"`

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,249 @@
+		capability "Refresh"
+		capability "Switch"
+		capability "Switch Level"
+		capability "Light"
```
This capability is outdated, you can remove it

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,249 @@
+			} else {
+				log.warn "ON/OFF REPORTING CONFIG FAILED- error code:${cluster.data[0]}"
+			}
+		} else {}
```
empty block like this `else {}` is definitely not needed

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,249 @@
+	if (event) {
+		sendEvent(event)
+	} else if (description?.startsWith('read attr -')) {
+		if (zigbeeMap.cluster == "0202" &&
```
Please change all "magic numbers" to constants, like:
`private getBATTERY_PERCENTAGE_REMAINING() { 0x0021 }`
to which later you can refer as
`BATTERY_PERCENTAGE_REMAINING`
(this particular example is from https://github.com/SmartThingsCommunity/SmartThingsPublic/blob/master/devicetypes/smartthings/leaksmart-water-sensor.src/leaksmart-water-sensor.groovy#L62 )

Later, you can compare them by using `zigbeeMap.clusterInt` instead of `zigbeeMap.cluster`.
Example of usage: https://github.com/SmartThingsCommunity/SmartThingsPublic/blob/a7c980340992ca37508d42a70f354b5fd7d30048/devicetypes/smartthings/zigbee-power-meter.src/zigbee-power-meter.groovy#L58

PKacprowiczS replied:
This applies to all of "magic numbers", not only this one.

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,82 @@
+        	sendEvent(name: "switch", value: "off", displayed: true, isStateChange: true)
+    	} else {
+        	sendEvent(name: "switch", value: "on", displayed: true, isStateChange: true)
+    	}
```
Don't create events as a part of performing a command. Send those events back, after parent device receives communication confirming that command was successful.

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,249 @@
+}
+
+def addChildFan() {
+        log.debug "add child fan"
```
nit: spacing

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,249 @@
+		capability "Health Check"
+		capability "Refresh"
+		capability "Switch"
+		capability "Switch Level"
```
Why do use this capability as a "bridge" to a Fan Speed, when you can just use `capability "Fan Speed"`? 
Using 0-100 range to control 5 speeds doesn't seem to be a great UX approach.

RaihaPark replied:
Because we want connect other platforms via SmartThings cloud to cloud connection. But other platforms don't support "Fan speed" capability. So we have to use "Switch level" capability for control fan speed.

PKacprowiczS replied:
I don't like it, but if it has to be this way, so be it. Would it be a problem for this "other platform" if you'd redefine range of `Switch Level` capability to be 0-4, instead of 0-100?

RaihaPark replied:
I'm fully agree with you. I don't like it too. Other platforms only support switch level range 0 to 100. I can't change the level range. It is absurd but there is no way.

RaihaPark committed:
Update led-fan-lightings.groovy

Edit code style

RaihaPark committed:
Update itm-fan-child.groovy

Edit code style

RaihaPark committed:
Update led-fan-lightings.groovy

change code style

RaihaPark committed:
Update led-fan-lightings.groovy

Change code style

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,242 @@
+	if (event) {
+		sendEvent(event)
+	} else if (description?.startsWith('read attr -')) {
+		if (zigbeeMap.cluster == FAN_CLUSTER_VALUE &&
```
@RaihaPark please double check if this comparison have same types on both sides, because as far as I can remember, `parseDescriptionAsMap(desc).cluster` is a string, not an integer and perhaps you should use `clusterInt`

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -20,15 +20,10 @@ import physicalgraph.zigbee.zcl.DataType
-		capability "Refresh"
-		capability "Switch"
-		capability "Switch Level"
-		capability "Light"
```
I was referring only to `capability "Light"`.

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -20,15 +20,10 @@ import physicalgraph.zigbee.zcl.DataType
 import groovy.json.JsonOutput
 
 metadata {
-	definition (name: "LED FAN lightings", namespace: "SAMSUNG LED", author: "SAMSUNG LED", runLocally: true, minHubCoreVersion: '000.019.00012', executeCommandsLocally: true, genericHandler: "Zigbee") {
```
You forgot about `genericHandler: "Zigbee"`

RaihaPark replied:
Thank you so much :) I fixed it follow your advice.

RaihaPark committed:
Merge branch 'SmartThingsCommunity:master' into master

RaihaPark committed:
Update led-fan-lightings.groovy

Update code style

RaihaPark committed:
Update led-fan-lightings.groovy

Edit code style

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -67,8 +67,8 @@ def parse(String description) {
 	if (event) {
 		sendEvent(event)
 	} else if (description?.startsWith('read attr -')) {
-		if (zigbeeMap.cluster == FAN_CLUSTER_VALUE &&
```
Don't bring back magic numbers. As I've already wrote you, you can use `zigbeeMap.clusterInt` to correctly compare number values.
This is the part of `parseDescriptionAsMap(description)` method which proves it. 
```
if (outMap.containsKey("cluster")) {
	outMap["clusterInt"] = convertHexToInt(outMap["cluster"])
}
if (outMap.containsKey("clusterId")) {
	outMap["clusterInt"] = convertHexToInt(outMap["clusterId"])
}
if (outMap.containsKey("attrId")) {
	outMap["attrInt"] = convertHexToInt(outMap["attrId"])
}
if (outMap.containsKey("command")) {
	outMap["commandInt"] = convertHexToInt(outMap["command"])
}
```

RaihaPark replied:
Are those integers hexadecimal too?

PKacprowiczS replied:
Number base actually doesn't matter, you can use both, for example `0x0202` and `514`, but obviously it's better to use first (hex) option in code, as it relates to Zigbee documentation.

PKacprowiczS replied:
To extend my response, `0x0202 == 514` would be `true` in Groovy.

RaihaPark replied:
Oh, thank you for explain. I understand.

RaihaPark committed:
Update led-fan-lightings.groovy

Change code style

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,245 @@
+	}	
+	try {
+		String dni = "${device.deviceNetworkId}-Fan"
+		addChildDevice("ITM Fan Child", dni, device.hub.id, [completedSetup: true, label: "${componentLabel}", isComponent: false])
```
I'd prefer if you'd set DNI to `"${device.deviceNetworkId}:1"`.

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,245 @@
+
+def setLevel(value, rate = null, device = null) {
+	if (device == null) {
+		zigbee.setLevel(value)
```
Does this part of `setLevel` relates to Light which can be dimmed? Cause using `zigbee.setLevel(value)` will send a command to LEVEL_CONTROL_CLUSTER.

RaihaPark replied:
Oh, sure. It is the most important part of this code. For separate operation between dimming and fan speed "device" added for it. Because of the separator "device",  "LEVEL_CONTROL_CLUSTER" can operates separately.

RaihaPark committed:
Update led-fan-lightings.groovy

Change point (4 codes)
"${device.deviceNetworkId}-Fan" to "${device.deviceNetworkId}:1"

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,245 @@
+            		setFanSpeed(3, device)
+        	} else if (value <= 100) {
+            		setFanSpeed(4, device)
+        	}        
```
So referring to discussion here https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/68445/files#r670393336, will this part of code be actually ever reached? It isn't called from the child device, and I highly doubt that argument `device` will be ever different than `null`, as it even not considered as a capability's command argument.

RaihaPark replied:
That code is a function overloading. It can separates what is called between a child or a main.

PKacprowiczS replied:
Walk me through it, cause I don't see it. Child DTH calls only parent's `setFanSpeed`.

RaihaPark replied:
What a cool programmer you are! Yes, you're right. Thanks for looking closely. This part uses for cloud to cloud connection. I told before that there is not fan speed capability in other platforms. So we have to use switch level capability instead of fan speed capability. Then other platform user can see the switch level UI for control fan speed. When the user change the switch level for fan speed, child DTH calls parent's setFanSpeed. And that point of the time SmartThings hub operates for fan speed control.

PKacprowiczS replied:
Are you able to tell me which 3rd party cloud you want to integrate it with? 
Will it call SmartThings API? 
Have you tested communication between ST and 3rd party with this current DTH?
Right now I doubt it'd work, cause `device` isn't listed as command's attribute https://smartthings.developer.samsung.com/docs/api-ref/capabilities.html#Switch-Level

RaihaPark replied:
![image](https://user-images.githubusercontent.com/74279632/127978037-cdad118b-8c02-4880-85fd-f6c397bb2443.png)

RaihaPark replied:
Sorry for late answer. I needed to change code to use capability "Switch Level" for control fan speed in other platforms. I upload an image file. I captured some screens when I use different platforms. You can see that. Those are google home and amazon alexa.

PKacprowiczS replied:
Move this logic to to `setLevel()` you've created in the child DTH.

RaihaPark committed:
Update itm-fan-child.groovy

Change code
- Add capability "Switch Level"
- Add function setLevel

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,85 @@
+
+def setLevel(value){
+	parent.setLevel(value, null, device)
+}
```
Wouldn't it be better to call `setFanSpeed()` (preferably the one from the child DTH)? In that case you'd hold `Light` and `Fan speed` logic apart (and save some headache to anyone who would try to figure out what's going on).

RaihaPark replied:
OK, I changed it :)

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,85 @@
+        	capability "Configuration"
+        	capability "Refresh"
+        	capability "Switch"
+		capability "Switch Level"
```
Can you add some comment about `Switch Level` capability, that it's really needed for controlling fans in 3rd party APIs.

RaihaPark replied:
Sure, If I don't add Switch Level capability, level control bar doesn't appears. Only switch button shows.

PKacprowiczS replied:
I've meant including comment in the code.

RaihaPark replied:
Got it! Thank you

RaihaPark committed:
Update led-fan-lightings.groovy

Change code style

RaihaPark committed:
Update itm-fan-child.groovy

Change code style

RaihaPark committed:
Merge branch 'SmartThingsCommunity:master' into master

RaihaPark committed:
Update itm-fan-child.groovy

Add a comment for capability "Switch Level"

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,97 @@
+	} else if (value <= 100) {
+		setFanSpeed(4)
+       	}	
+}
```
Fix up indentations in the method. And remember about lacking space between `)` and `{` in line 43.

RaihaPark replied:
Oh. I missed it. Thank you.

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,231 @@
+		if (zigbeeMap.clusterInt == FAN_CLUSTER_VALUE &&
+		    zigbeeMap.attrInt == FAN_STATUS_VALUE) {		
+			log.debug "read attribute event for fan cluster attrib FAN_STATUS"
+			def childDevice = getChildDevices()?.find {
```
`getChildDevices()` can be also adressed as `childDevices` and it's preferred way

RaihaPark replied:
Got it! Thank you

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,231 @@
+	// Parse incoming device messages to generate events
+	log.debug "description is $description"
+	def event = zigbee.getEvent(description)
+	def zigbeeMap = zigbee.parseDescriptionAsMap(description)
```
I'd move this call to `else if (description?.startsWith('read attr -')) {...}` block.

RaihaPark replied:
The line 66 moved to that block.

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,231 @@
+        	log.debug "max"
+	    	sendFanSpeed(0x04)
+    	}    
+}
```
Is this necessary to pass `device` to this method (obviously assuming that you also modify calls in the child device to `parent.setFanSpeed(speed)`? It's not used anywhere, device has only one child device, the method is called only from the child device.

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,231 @@
+	zigbee.onOffRefresh() +
+	zigbee.levelRefresh() +
+	zigbee.onOffConfig(0, 300) +
+	zigbee.levelConfig()+
```
I don't think configs are necessary in the `refresh()`.

PKacprowiczS replied:
But you might want to include them in `configure()`.

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,231 @@
+			deleteChildren()
+		}
+    	} catch(e) {}
+}
```
Children are deleted automatically, you don't need this part of code.

RaihaPark replied:
Thank you a lot. Codes changed much more slim and clear.

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,231 @@
+    	} else {		
+        	log.debug "max"
+	    	sendFanSpeed(0x04)
+    	}    
```
If you'd concatenated it to one, `else {...}` block, wouldn't it be the same?

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,97 @@
+    	} else {
+        	sendEvent(name: "switch", value: "on", displayed: true, isStateChange: true)
+    	}	
+	parent.setFanSpeed(speed, device)
```
Another thought - what do you think if you'd just call parent's `sendFanSpeed()` here, instead of `setFanSpeed()`? Basically, you're just passing the same numbers there.

PKacprowiczS replied:
@RaihaPark have you thought about this?

PKacprowiczS replied:
@RaihaPark I'd really prefer if you'd address it. At least explain why do you think it's not a good idea to change it.
By "it", I mean my comment I made in previous review https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/68445#discussion_r682591717

RaihaPark replied:
Oh, I'm sorry. I missed it. But we designed this code parents side have all control functions. So there is a parse function only in the parents side. Because of other platform issue we have to did that, especially alexa. If we call directly fanspeed function in the child side as like line 146, the fan controls well but we see some error messages together. So we had to separate control functions to parent side.

PKacprowiczS replied:
I think we don't understand each other. 
Let's assume there's a UC - user sets fan speed to 75% through Alexa.
Current flow is:
1. Alexa calls, with SmartThings API, child device's `Switch Level`'s command `setLevel(value)` with `value = 75`.
2. Child DTH calls `setFanSpeed(speed)` with `speed = 3`. 
3. We're still in Child DTH - now parent's method is called - `parent.setFanSpeed(speed, device)`
4. We're in parent DTH, in `setFanSpeed(value, device)` method, where `value = 3` and `sendFanSpeed(val)` is called.
5. In `sendFanSpeed(val)`, finally, a zigbee command is send.

If you've done what I want you to do, it'll be reduced to:
1. Alexa calls, with SmartThings API, child device's `Switch Level`'s command `setLevel(value)` with `value = 75`.
2. Child DTH calls `setFanSpeed(speed)` with `speed = 3`. 
3. Parent's method is called - `parent.sendFanSpeed(speed)`
4. We're in parent DTH, in `sendFanSpeed(val)` method, where `val = 3` and a zigbee command is send.

You remove one extra step, but what's more important you clean up the code big time. 
In parent DTH you really don't need `setFanSpeed()`, it's only called from child DTH as parent doesn't have `Fan Speed` now.
All you need to do is remove `setFanSpeed(speed, device)` from parent DTH and change 64 line in child DTH from
```
parent.setFanSpeed(speed, device)
```
to
```
parent.sendFanSpeed(speed)
```
and you'd have same functionality with less pain.

RaihaPark replied:
Oh, I see. You're right. I'll clean up the codes.

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,97 @@
+        	sendEvent(name: "switch", value: "off", displayed: true, isStateChange: true)
+    	} else {
+        	sendEvent(name: "switch", value: "on", displayed: true, isStateChange: true)
+    	}	
```
I'd prefer if those events were created as a response of the command, not automatically when command is triggered.

PKacprowiczS replied:
@RaihaPark what about this?

PKacprowiczS replied:
@RaihaPark I'd really prefer if you'd address it. At least explain why do you think it's not a good idea to change it.
By "it", I mean my comment I made in previous review https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/68445#discussion_r682592329

PKacprowiczS replied:
@RaihaPark I guess you've missed this conversation.

RaihaPark replied:
Oh, I'm sorry. I missed it. But sometimes when I change the swithlevel, switch does not change to "on" or "off" in other platforms. So I put this code in.

PKacprowiczS replied:
Maybe it's caused by too low delay here? https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/68445/files#diff-0983345aeab4215b6a7f7c49ebfe0e649324ac2d64018e6c8bffe0c9b24692b6R123

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,97 @@
+
+def parse(String description) {
+	log.debug "[Child] - PARSE IN Child: $description"
+}
```
This won't be ever called, you might as well delete it.

RaihaPark replied:
Thank you a lot. Codes changed much more slim and clear.

RaihaPark committed:
Update itm-fan-child.groovy

Code update

RaihaPark committed:
Update led-fan-lightings.groovy

Code update

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,93 @@
+	} else if (value <= 100) {
+		setFanSpeed(4)
+	}	
+}
```
Yeah, it'd be great to align whole method, not only first & last line.

RaihaPark replied:
Thanks a lot. I missed it. :)

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,206 @@
+def installed() {
+	log.debug "Samsung ITM test prod installed"
+	addChildFan()    
+	configure()
```
No need to call `configure()`, since you've declared `Configuration` capability.

RaihaPark replied:
Oh, I didn't know that. Thank you.

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -62,15 +62,15 @@ private getON_OFF_CLUSTER_VALUE() { 0x0006 }
 		    zigbeeMap.attrInt == FAN_STATUS_VALUE) {		
 			log.debug "read attribute event for fan cluster attrib FAN_STATUS"
-			def childDevice = getChildDevices()?.find {
+			def childDevice = childDevice.find {
```
`getChildDevices()` => `childDevices`, not `childDevice`. Plural, not singular.

RaihaPark replied:
Thank you very much. By your review the codes go to more accurate.

PKacprowiczS reviewed (changes requested):
Please also have a look at those remaining comments from the last review.

RaihaPark committed:
Update itm-fan-child.groovy

Align codes.

RaihaPark committed:
Update led-fan-lightings.groovy

Edit codes

RaihaPark committed:
Update led-fan-lightings.groovy

Clean up the codes

RaihaPark committed:
Update itm-fan-child.groovy

Clean up the codes

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,93 @@
+
+def on() {
+	setFanSpeed(1)
+}
```
It's a minor thing, but to keep consistency, I'd change it to
```
def off() {
	setFanSpeed(0x00)
}

def on() {
	setFanSpeed(0x01)
}
```

RaihaPark replied:
Oh, Thank you :) Got it!

PKacprowiczS reviewed (changes requested):
Almost done.

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,182 @@
+			}          
+			event.displayed = true
+			zigbeeMap.name = "fanSpeed"
+			childDevice.sendEvent(zigbeeMap)
```
Why are you sending `zigbeeMap` as an event to child DTH? This should a proper event map/object.

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,182 @@
+				int_v = int_v * 25
+				int_v = int_v > 100 ? 100 : int_v
+				log.debug "child device Level set $int_v"
+				childDevice.sendEvent(name: "level", value: int_v)	// For cloud to cloud device UI update
```
Shouldn't you, also, send here a proper `fanSpeed` events?

RaihaPark replied:
That is for the UI control when it has made cloud to cloud connection. So, it uses "switch" capability and "switch level" capability instead of "fanspeed" capability for UI update.

PKacprowiczS replied:
Yeah, I get it. I clearly meant that you perhaps should send a `fanSpeed` event in addition, not instead.

PKacprowiczS replied:
Because of what I was addressing here https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/68445/files#r686664824 I believe there is no correct `fanSpeed` event send as a response to a Zigbee command.

RaihaPark replied:
The line 79 zigbeeMap.name = "fanSpeed" means designate a child UI event name. And the line 89 to 90 means matching the fan speed to switch level.

PKacprowiczS replied:
Even though it's working, cause `zigbeeMap` indeed contains key `value`, I'd like you to create a separate map object, like `fanSpeedEvent` and send it instead. This way it'll be more clear what are you doing.

RaihaPark replied:
Oh, great! Not for cause the confusion I'll change the name.

PKacprowiczS reviewed (changes requested):
Another things I've missed.

RaihaPark committed:
Update itm-fan-child.groovy

Fix the codes

RaihaPark committed:
Update itm-fan-child.groovy

Clean up the codes

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,182 @@
+				log.debug "parse() child device found"
+				it.device.deviceNetworkId == "${device.deviceNetworkId}:1" 
+			}          
+			event.displayed = true
```
This statement doesn't have any effect, as you're not doing anything with `event` later.

RaihaPark committed:
Update led-fan-lightings.groovy

Edit the name of zigbeeMap.name

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -76,7 +76,7 @@ def parse(String description) {
 			}          
 			event.displayed = true
-			zigbeeMap.name = "fanSpeed"
+			zigbeeMap.name = "fanSpeedEvent"
```
I've meant more like
```
def fanSpeedEvent = createEvent(name: "fanSpeed", value: zigbeeMap.value)
child.sendEvent(fanSpeedEvent)
```
The way you changed it, it won't work at all. And remember also about looking at this comment - https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/68445#discussion_r687482840.

RaihaPark replied:
Thank you very much :) I fixed it.

RaihaPark committed:
Update led-fan-lightings.groovy

Edit codes

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,181 @@
+				log.debug "parse() child device found"
+				it.device.deviceNetworkId == "${device.deviceNetworkId}:1" 
+			}
+			def fanSpeedEvent = createEvent(name: "fanSpeed", value: zigbeeMap.value)
```
Ok, so please just verify whether it shouldn't be `value: zigbeeMap.value as Integer`, because `zigbeeMap.value` in this case is a string, and `fanSpeed` attribute accepts integers, and that'd be all from me.

RaihaPark replied:
Ok, I changed it.

PKacprowiczS reviewed (dismissed):
@greens do you have a moment to have a look at this PR?

RaihaPark committed:
Update led-fan-lightings.groovy

Edit code

RaihaPark committed:
Update led-fan-lightings.groovy

Edit codes

greens commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,89 @@
+/*
+ *  Copyright 2017 SmartThings
```
2021

greens commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,181 @@
+/*
+ *  Copyright 2017 SmartThings
```
2021

greens commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,181 @@
+	} else {
+		def cluster = zigbee.parse(description)
+		if (cluster && 
+		    cluster.clusterId == ON_OFF_CLUSTER_VALUE &&
```
should be `clusterInt`

greens commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,181 @@
+	delayBetween([zigbee.writeAttribute(FAN_CLUSTER_VALUE, FAN_STATUS_VALUE, DataType.ENUM8, val), zigbee.readAttribute(FAN_CLUSTER_VALUE, FAN_STATUS_VALUE)], 100)
+}
+
+def ping() {
```
child DTH calls this with arguments

greens replied:
and if onOffRefresh doesn't return a value that generates an event for the child anyway, then it will not keep the device marked online

greens commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,181 @@
+	zigbee.setLevel(value)
+}
+
+private sendFanSpeed(val) {
```
child DTH won't be able to access this method

PKacprowiczS replied:
`private` -> `def`

RaihaPark committed:
Update itm-fan-child.groovy

Edit year

RaihaPark committed:
Update led-fan-lightings.groovy

Edit year

RaihaPark committed:
Update led-fan-lightings.groovy

Edit code

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,89 @@
+
+void refresh() {
+	log.debug "[Child] - refresh()"
+	parent.refresh(device)
```
Similar thing as Steven has noticed for `ping()`. You're calling `refresh` with `device` as argument, but there's no refresh which would accept any argument (and it doesn't need it, really).

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,89 @@
+void refresh() {
+	log.debug "[Child] - refresh()"
+	parent.refresh(device)
+	parent.childRefresh(device.deviceNetworkId)
```
`childRefresh` method doesn't exist in parent DTH.

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,89 @@
+		/* Capability "Switch Level" is used to control fan speed for platforms don't support capability "Fan speed"
+		 * when you connect other platforms via SmartThings cloud to cloud connection. */
+		capability "Switch Level"
+        	capability "Fan Speed"
```
Align these lines.

PKacprowiczS replied:
@RaihaPark, I've meant all the lines - from 24 to 31.

RaihaPark replied:
Ok

PKacprowiczS replied:
@RaihaPark - align all lines - make them even, make them start at the same point. Not move to the left all of them. Do you not really see that lines 25, 26 and 31 have extra spaces compared to the rest?

RaihaPark replied:
I'm sorry. I didn't saw that.

RaihaPark committed:
Update itm-fan-child.groovy

Edit code

RaihaPark committed:
Update itm-fan-child.groovy

Edit code

RaihaPark committed:
Update itm-fan-child.groovy

Align codes

RaihaPark committed:
Update led-fan-lightings.groovy

Align codes

RaihaPark committed:
Update itm-fan-child.groovy

Align codes

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,181 @@
+def ping() {
+	// PING is used by Device-Watch in attempt to reach the Device
+	return zigbee.onOffRefresh()
+}
```
As Steven has said, you also need this method to return a command that would cause sending an event for child device, e.g. `zigbee.readAttribute(FAN_CLUSTER_VALUE, FAN_STATUS_VALUE)`

RaihaPark replied:
OK. I see.

RaihaPark committed:
Update led-fan-lightings.groovy

Align codes

RaihaPark committed:
Update itm-fan-child.groovy

Align code

RaihaPark committed:
Update led-fan-lightings.groovy

Edit code

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -124,6 +124,7 @@ private sendFanSpeed(val) {
 
 def ping() {
 	// PING is used by Device-Watch in attempt to reach the Device
+	zigbee.readAttribute(FAN_CLUSTER_VALUE, FAN_STATUS_VALUE)
```
This won't do anything, as it isn't returned from the method.

You need something like
```
def ping() {
	return zigbee.onOffRefresh() +
		zigbee.readAttribute(FAN_CLUSTER_VALUE, FAN_STATUS_VALUE)
}
```
or
```
def ping() {
	zigbee.onOffRefresh() +
	zigbee.readAttribute(FAN_CLUSTER_VALUE, FAN_STATUS_VALUE)
}
```

RaihaPark replied:
Thank you so much. I really appreciate it.

RaihaPark committed:
Update led-fan-lightings.groovy

Fix codes

RaihaPark committed:
Update led-fan-lightings.groovy

Fix codes

RaihaPark committed:
Merge branch 'SmartThingsCommunity:master' into master

RaihaPark committed:
Merge branch 'SmartThingsCommunity:master' into master

greens commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -0,0 +1,87 @@
+}
+
+def configure() {
+	log.debug "[Child] - configure()"   
```
if you're done debugging, please remove all of these no-op methods that just log that they were called

RaihaPark replied:
Ah, OK. I'll do that :)

greens commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,182 @@
+	zigbee.on()
+}
+
+def setLevel(value) {
```
i believe the spec requires you also include a `duration` argument for this function, but you can simply include it as an argument but not use it

RaihaPark replied:
Do you mean "setLevel" needs some arguments for duration? And how can I fix it not to use?

PKacprowiczS replied:
`def setLevel(value) {` -> `def setLevel(value, duration) {`

RaihaPark replied:
Thanks a lot

greens commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,182 @@
+	// OnOff minReportTime 0 seconds, maxReportTime 5 min. Reporting interval if no activity
+	zigbee.onOffConfig(0, 300)
+	zigbee.levelConfig()
+	refresh()
```
these three must be concatenated together into a list

just calling `zigbee.onOffConfig` does not send the command, it must be returned by `configure`

RaihaPark replied:
I fixed those.

greens commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,182 @@
+	} catch(e) {
+		log.warn "Failed to add ITM Fan Controller - $e"
+	}
+    	def childDevice = getChildDevices()?.find {
```
the device is returned by `addChildDevice` so you can skip this step if you just set up this variable before that call

RaihaPark replied:
Can I skip the try and catch function?

PKacprowiczS replied:
I'd leave try & catch blocks.

RaihaPark replied:
I leave it as is.

RaihaPark committed:
Update led-fan-lightings.groovy

Delete log.bebug *

RaihaPark committed:
Update itm-fan-child.groovy

Delete log.debug *

RaihaPark committed:
Update led-fan-lightings.groovy

Fix the configure function

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -55,33 +55,26 @@ def setLevel(value) {
 
 def updated() {
-	log.debug "[Child] - updated()"
 }
```
Since those methods don't do anything, you can remove them.

PKacprowiczS replied:
I've meant both `configure()` and `updated()`, but GitHub has played a trick on me.

RaihaPark replied:
Then can I remove install(), too? It just calls configure().

PKacprowiczS replied:
Yeah, good idea.

RaihaPark replied:
I fixed it :)

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/itm-fan-child.groovy:
```diff
@@ -55,33 +55,26 @@ def setLevel(value) {
 
 def uninstalled() {
-	log.debug "[Child] - uninstalled()"	
 }
```
You can remove `uninstalled()` as well.

RaihaPark replied:
OK, I cleared them.

RaihaPark committed:
Update led-fan-lightings.groovy

Edit codes

RaihaPark committed:
Update itm-fan-child.groovy

Clear codes

RaihaPark committed:
Update itm-fan-child.groovy

Clear codes

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -105,8 +105,8 @@ def on() {
-def setLevel(value) {
-	zigbee.setLevel(value)
+def setLevel(value, duration) {
+	zigbee.setLevel(value, duration)
```
You don't have to call `duration` in `zigbee.setLevel`.

RaihaPark replied:
OK, I'll fix it

RaihaPark committed:
Update led-fan-lightings.groovy

Fix the code

RaihaPark committed:
Update itm-fan-child.groovy

Clear codes

RaihaPark committed:
Merge branch 'SmartThingsCommunity:master' into master

RaihaPark committed:
Update zigbee-rgbw-bulb.groovy

Add the finger print for ABL Juno connect E-series

RaihaPark committed:
Merge pull request #1 from RaihaPark/RaihaPark-ABL-E-series

WWST-5862, add the FP for ABL E-series

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,167 @@
+	if (childDevice != null) {
+		childDevice.sendEvent(name: "switch", value: "off")
+	}
+}
```
Change this to:
```
def addChildFan() {
	def componentLabel
	def childDevice

	if (device.displayName.endsWith(' Light') ||
	    device.displayName.endsWith(' light')) {
		componentLabel = "${device.displayName[0..-6]} Fan"
	} else {
		// no '1' at the end of deviceJoinName - use 2 to indicate second switch anyway
		componentLabel = "$device.displayName Fan"
	}	

	try {
		String dni = "${device.deviceNetworkId}:1"
		childDevice = addChildDevice("ITM Fan Child", dni, device.hub.id, [completedSetup: true, label: "${componentLabel}", isComponent: false])
	} catch(e) {
		log.warn "Failed to add ITM Fan Controller - $e"
	}

	if (childDevice != null) {
		childDevice.sendEvent(name: "switch", value: "off")
	}
}
```

RaihaPark replied:
Ah, OK! I change it. Thank you so much :)

PKacprowiczS reviewed (changes requested):
Sorry, I didn't notice that you've missed Steven's last suggestion.

RaihaPark committed:
Update led-fan-lightings.groovy

Edit addChildFan function

RaihaPark committed:
Merge branch 'SmartThingsCommunity:master' into master

PKacprowiczS commented on devicetypes/smartthings/zigbee-ceiling-fan-light.src/led-fan-lightings.groovy:
```diff
@@ -0,0 +1,166 @@
+}
+
+def delete() {
+}
```
Remove this empty method.

RaihaPark replied:
OK, I cleared that.

RaihaPark committed:
Update led-fan-lightings.groovy

Clear codes

RaihaPark committed:
Merge branch 'SmartThingsCommunity:master' into master

RaihaPark commented:
Thank you so much :)
Because of your sharp review, I studied more too.
Next time when I make other code, I can do better!

greens commented:
@RaihaPark please correct the title to reference the correct jira ticket

RaihaPark commented:
@greens Thank you so much :)
